✨ Cancel hook on rebase

Jeremie-Chauvel opened the pull request:
## Description
Hi @carloscuesta, I recreated the feature to cancel the hook on rebase.

Noteworthy point, I contacted @adriencaccia, which implemented a elegant solution on a fork to prefill the commit message in this case, he will come back to you with a proposition .

<!-- Explanation about your pull request, what changes you've made -->

<!-- Add issue number that this pull request refers to -->
Issue: https://github.com/carloscuesta/gitmoji-cli/issues/444

## Tests

<!-- Ensure that all the tests passed -->
- [x] All tests passed.

carloscuesta commented on test/setupTests.js:
```diff
@@ -12,7 +12,7 @@ jest.mock('ora', () =>
 jest.mock('conf')
 jest.mock('inquirer')
-jest.mock('execa')
+jest.mock('execa', () => jest.fn(() => Promise.resolve({ stdout: 'master' })))
```
I would leave this as a general mock like this:

```suggestion
jest.mock('execa')
```

And then mock on the tests as needed the parts of the module you'll need to mock

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -118,6 +118,7 @@ describe('commit command', () => {
         getEmojis.mockResolvedValue(stubs.gitmojis)
         mockProcess.mockProcessExit()
         process.argv[3] = stubs.argv
+        execa.mockReturnValueOnce(Promise.resolve({ stdout: 'master' }))
```
I think we should update the describe to match to the new spec

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -142,6 +143,7 @@ describe('commit command', () => {
         getEmojis.mockResolvedValue(stubs.gitmojis)
         mockProcess.mockProcessExit()
         process.argv[3] = stubs.argv
+        execa.mockReturnValueOnce(Promise.resolve({ stdout: 'master' }))
```
I think we should update the describe to match to the new spec

Jeremie-Chauvel replied:
like `while committing on master`?

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -157,9 +159,30 @@ describe('commit command', () => {
 
+    describe('when rebasing', () => {
+      it('should cancel the hook', async () => {
+        // Use an exception to suspend code execution to simulate process.exit
+        mockProcess.mockProcessExit(new Error('ProcessExit0'))
```
Why we can't simulate process.exit here ? 🤔

Jeremie-Chauvel replied:
I'm simulating process exit by using an exception, if process exit 1 happen during the test, the test is killed, that's why I mock the process exit. I'm using an exception to interrupt as well the code execution to further mimic a process exit

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -170,6 +193,8 @@ describe('commit command', () => {
           processEvents[signal]()
         })
 
+        execa.mockReturnValueOnce(Promise.resolve({ stdout: 'master' }))
```
We need to update the describe expect to match this new condition here

carloscuesta commented on src/commands/commit/withHook/index.js:
```diff
@@ -26,4 +27,13 @@ export const registerHookInterruptionHandler = () => {
 }
 
+export const cancelIfRebasing = (mode: 'hook' | 'client') =>
+  mode !== 'hook'
```
Why not making this simpler and then switch the ternary order

```suggestion
  mode === 'hook'
```

Jeremie-Chauvel replied:
ok for me

carloscuesta commented on src/commands/commit/withHook/index.js:
```diff
@@ -26,4 +27,13 @@ export const registerHookInterruptionHandler = () => {
 
+export const cancelIfRebasing = (mode: 'hook' | 'client') =>
+  mode !== 'hook'
+    ? Promise.resolve()
+    : execa('git', ['branch']).then(({ stdout: branchNames }) => {
+        if (/no branch/.test(branchNames)) {
+          process.exit(0)
+        }
+      })
```
```suggestion
    ? execa('git', ['branch']).then(({ stdout: branchNames }) => {
        if (/no branch/.test(branchNames)) {
          process.exit(0)
        }
      })
    : Promise.resolve() 
```

Jeremie-Chauvel replied:
ok for me

carloscuesta commented on src/commands/commit/withHook/index.js:
```diff
@@ -26,4 +27,13 @@ export const registerHookInterruptionHandler = () => {
   })
 }
 
+export const cancelIfRebasing = (mode: 'hook' | 'client') =>
```
I think we can make this function `async` to benefit from this and be able to use `await` keyword inside of the fn

Jeremie-Chauvel replied:
see https://github.com/carloscuesta/gitmoji-cli/pull/445#discussion_r477316501

carloscuesta commented on src/commands/commit/withHook/index.js:
```diff
@@ -26,4 +27,13 @@ export const registerHookInterruptionHandler = () => {
+export const cancelIfRebasing = (mode: 'hook' | 'client') =>
+  mode !== 'hook'
+    ? Promise.resolve()
+    : execa('git', ['branch']).then(({ stdout: branchNames }) => {
+        if (/no branch/.test(branchNames)) {
+          process.exit(0)
+        }
+      })
```
Can you explain why you're looking to branch names ? That's why because on the `rebase` process you're not on a named branch but on a commit revision ? There's another way to check if we're on a rebase process ?

Jeremie-Chauvel replied:
Hi exactly during a rebase, the current branch is `no branch, rebasing <branch name>`, I suppose using git status could do as well

carloscuesta commented on src/commands/commit/index.js:
```diff
@@ -3,21 +3,28 @@ import inquirer from 'inquirer'
+  }
 
-  return getEmojis()
-    .then((gitmojis) => prompts(gitmojis))
-    .then((questions) => {
-      inquirer.prompt(questions).then((answers) => {
-        if (mode === 'hook') return withHook(answers)
+  return cancelIfRebasing(mode).then(() =>
```
There's something I think we can improve here.

If we're already making a conditional for the `mode` inside of this function maybe we can reuse it and execute the `cancelIfRebasing` function here

Jeremie-Chauvel replied:
Sadly, it's not possible to use a function with await in if(mode) since then:
- We need to amke commit an async func
- Then in 'client' mode, we have the 'hang' that you observed in hook mode (the cli doesn't complete until ctrl-C)

carloscuesta reviewed (commented):
Thanks! A small details about the code ❤️

carloscuesta commented:
I've been testing out this branch locally linked and I found something weird:

When I try to make a commit with the hook option and I finish the prompts. The gitmoji-cli process does not ends until I hit CTRL+C

We should fix this behaviour before merging this:

Demo attached

![2020-08-25 15-35-24 2020-08-25 15_36_54](https://user-images.githubusercontent.com/7629661/91181106-d1824c80-e6e8-11ea-85f6-e19c7f2572a0.gif)

To test your changes you'll need to uninstall the cli from your system globally:

````
npm uninstall -g gitmoji-cli
```

Clone the project, checkout to your branch and run 

```
npm run build && npm run link
```

Jeremie-Chauvel commented:
Hi thanks for the feedback,
concerning the hang, I pretty sure I fixed it, I will look into it again.

Jeremie-Chauvel commented:
Hi @carloscuesta
I addressed your feedback:
- I found a way to use the branch mode for this new is rebase test
- I also used a different way to detect rebase, using a git folder created during a rebase see https://stackoverflow.com/questions/3921409/how-to-know-if-there-is-a-git-rebase-in-progress for more details.

## client mode:
![2020-08-31-22-17-12](https://user-images.githubusercontent.com/28385941/91764609-e7f04280-ebd7-11ea-8dd3-dc6c9b33098f.gif)

## hook out of rebase:
![2020-08-31-22-19-13](https://user-images.githubusercontent.com/28385941/91764963-811f5900-ebd8-11ea-9140-4138bbffb795.gif)

## hook during rebase:
![2020-08-31-22-20-54](https://user-images.githubusercontent.com/28385941/91764925-6baa2f00-ebd8-11ea-8471-631981aac218.gif)

Jeremie-Chauvel commented:
hmm, concerning you freeze, it's caused by using client mode with hook registred, I'm looking into it, as I can reproduce it

Jeremie-Chauvel commented:
so, I added a fix for this as well, I tracked it down to a missing encoding in `isHookCreated
![2020-08-31-22-55-05](https://user-images.githubusercontent.com/28385941/91767914-2c321180-ebdd-11ea-849b-b7dc2feb2f23.gif)
`

Jeremie-Chauvel commented:
@carloscuesta could you take a look when you have some free time ?
Would be much appreciated, thanks.

carloscuesta commented:
Sure I will take a look at the PR <3

carloscuesta commented:
Hello @Jeremie-Chauvel 

I found the freeze cause and it's not related with your changes!

I'm going to put up a PR to fix it. It something introduced in another PR.

Afrer that I will review the PR

Thanks for the patience ❤️

carloscuesta commented:
This PR https://github.com/carloscuesta/gitmoji-cli/pull/461 fixes the "freeze" thing

carloscuesta commented on src/utils/isHookCreated.js:
```diff
@@ -11,7 +11,7 @@ const isHookCreated = async () => {
       return false
     }
-    return fs.readFileSync(stdout + HOOK.PATH) === HOOK.CONTENTS
+    return fs.readFileSync(stdout + HOOK.PATH, 'utf8') === HOOK.CONTENTS
```
I fixed this aswell! So when you rebase it there's no need to change that anymore! 🎉

carloscuesta commented on src/commands/commit/index.js:
```diff
@@ -3,13 +3,23 @@ import inquirer from 'inquirer'
+  return promptAndCommit(mode)
+}
+
+const promptAndCommit = (mode: 'client' | 'hook') =>
```
Can you create a `type CommitMode` for the `client` | `hook` thing ? Then we wont have to repeat those values!

carloscuesta commented on test/utils/isHookCreated.spec.js:
```diff
@@ -27,7 +27,8 @@ describe('isHookCreated', () => {
     expect(await isHookCreated()).toEqual(false)
     expect(fs.readFileSync).toHaveBeenCalledWith(
-      stubs.gitAbsoluteDir + HOOK.PATH
+      stubs.gitAbsoluteDir + HOOK.PATH,
+      'utf8'
```
Not needed anymore!

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -142,6 +144,8 @@ describe('commit command', () => {
         getEmojis.mockResolvedValue(stubs.gitmojis)
         mockProcess.mockProcessExit()
         process.argv[3] = stubs.argv
+        execa.mockReturnValueOnce(Promise.resolve(stubs.gitAbsoluteDir))
+        fs.existsSync.mockReturnValueOnce(false).mockReturnValueOnce(false)
```
Why we need to add those mocks inside this test ?

Jeremie-Chauvel replied:
to pass the is `cancelIfRebasing` test

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -118,6 +118,8 @@ describe('commit command', () => {
         getEmojis.mockResolvedValue(stubs.gitmojis)
         mockProcess.mockProcessExit()
         process.argv[3] = stubs.argv
+        execa.mockReturnValueOnce(Promise.resolve(stubs.gitAbsoluteDir))
+        fs.existsSync.mockReturnValueOnce(false).mockReturnValueOnce(false)
```
Why we need to add those mocks inside this test ?

Jeremie-Chauvel replied:
to pass the is `cancelIfRebasing` test

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -170,6 +191,9 @@ describe('commit command', () => {
           processEvents[signal]()
         })
 
+        execa.mockReturnValueOnce(Promise.resolve(stubs.gitAbsoluteDir))
+        fs.existsSync.mockReturnValueOnce(false).mockReturnValueOnce(false)
```
Why we need to add those mocks inside this test ?

Jeremie-Chauvel replied:
to pass the is `cancelIfRebasing` test

carloscuesta commented on test/commands/commit.spec.js:
```diff
@@ -157,9 +161,26 @@ describe('commit command', () => {
     })
 
+    describe('when rebasing', () => {
+      it('should cancel the hook', async () => {
+        // Use an exception to suspend code execution to simulate process.exit
+        mockProcess.mockProcessExit(new Error('ProcessExit0'))
```
It would be cool to "simulate" the "rebase" to test the `cancelIfRebasing` code 😊

Jeremie-Chauvel replied:
that's more or less what I'm doing when I mock ` fs.existsSync` to simulate detection of rebase

carloscuesta reviewed (commented):
I really like this solution!

Is much more simple and easy to follow ❤️ 

Thanks for the work you're putting in this PR we are really close to ship it!

carloscuesta reviewed (approved):
Let's ship this!

Thanks 🎉
